sympy/combinatorics/fp_groups.py

```diff
@@ -297,33 +297,66 @@ def __init__(self, G, gens):
         super(FpSubgroup,self).__init__()
         self.parent = G
         self.generators = list(set([g for g in gens if g != G.identity]))
+        self._min_words = None #for use in __contains__
         self.C = None
 
     def __contains__(self, g):
-        gens = [s for s in g.contains_generators()]
-        if any([s not in self.generators for s in gens]):
+        if self._min_words is None:
+            gens = self.generators[:]
+            gens.extend([e**-1 for e in gens])
+            for i, w1 in enumerate(gens):
+                for w2 in gens[i+1:]:
+                    if w2**-1 == w1:
+                        continue
+                    if w2[len(w2)-1]**-1 == w1[0] and w2*w1 not in gens:
+                        gens.append(w2*w1)
```

Comment:
Oh I think I see. I've checked in terminal and appending to the list while looping over it with `enumerate` does work properly, say:
```
>>> a = [1,2,3]
>>> for i, e in enumerate(a):
...    print(e)
...    if i == 2:
...        a.append(4)
1
2
3
4
```
There is potentially a different problem: the words appended in the loop won't be checked against the preceding words. E.g. if `gens == [b, b**-1, b*a*b**2]` and I add `b*a*b` and `a*b**2` to the end of the list, they won't be compared with `b` or `b**-1` once we get to them.